accepted/2022/FunctionPointers.md

```diff
@@ -0,0 +1,485 @@
+- [CallConvCdecl](https://docs.microsoft.com/dotnet/api/system.runtime.compilerservices.callconvcdecl)
+- [CallConvStdCall](https://docs.microsoft.com/dotnet/api/system.runtime.compilerservices.callconvstdcall)
+- [CallConvThiscall](https://docs.microsoft.com/dotnet/api/system.runtime.compilerservices.callconvthiscall)
+- [CallConvFastcall](https://docs.microsoft.com/dotnet/api/system.runtime.compilerservices.callconvfastcall)
+- [CallConvSuppressGCTransition](https://docs.microsoft.com/dotnet/api/system.runtime.compilerservices.callconvsuppressgctransition)
+- [CallConvMemberFunction](https://docs.microsoft.com/dotnet/api/system.runtime.compilerservices.callconvmemberfunction)
+
+C# in particular just maps calling conventions specified in a function pointer to `System.Runtime.InteropServices.CallConv*` types as metadata. These CallConv* types must live in the same assembly as defines `System.Object` (i.e. `System.Private.Corelib.dll`).
+
+_Off-topic thoughts on current implementation_: using these CallConv* classes to represent calling conventions was added in V6 because new callling conventions needed to be added and would not fit within the 4 bits that were being used to encode the existing calling conventions based on ECMA-335. So the decision, arguably not the best, was made to encode new calling conventions using custom modifiers since that would not affect existing metadata readers or require a ECMA-335 change. However, there are alternative approaches which may or may not have been considered -- one such approach would be to change the encompassing byte to be a 32-bit compressed integer encoding instead which .NET/CLI uses extensively elsewhere. This compressed integer approach is possible since the encompassing byte did not use bit 7 (`0x80`) which is the flag for a compressed integer to switch from one byte to two bytes. This would require a ECMA-335 change which would break other metadata readers once two+ bytes are used.
+
+### Exposing custom modifiers on `FieldInfo`, `ParameterInfo` and `PropertyInfo`
+Although we don't want to expose custom modifiers from function pointer types that are not calling conventions, obtaining the other custom modifiers is useful for those who read metadata for varying reasons. The [`MetadataLoadContext`](https://learn.microsoft.com/dotnet/api/system.reflection.metadataloadcontext) class is the preferred approach to read metadata, since it supports reading any assembly without having to load the respective types into the runtime.
+
+The proposal is to expose functionality in the `MetadataLoadContext` assembly using type extensions that return "signature types" which are an extended function pointer type that has `GetRequiredCustomModifiers()` and `GetOptionalCustomModifiers()` methods which return all custom modifiers. These APIs only work with signature types that are returned from `FieldInfo`, `PropertyInfo` and `ParameterInfo` via `GetFunctionPointerSignatureType()`.
```

Comment:
The reflection object model has a thing called signature type already: https://learn.microsoft.com/th-th/dotnet/api/system.type.issignaturetype

Are the signature types that we are talking about here same kind of signature types (ie `IsSignatureType` returns `true`) or are they something else? If they are something else, we should invent a new name for them.

Reply:
They are not the same -- I renamed them to "modified types".

For `IsSignatureType` see https://github.com/dotnet/runtime/issues/20377 which was added to make looking up generic methods from reflection easier. Also, I added `Type.CreateFunctionPointerType(Type[])` because lookup should also be possible based on a function pointer type.